### PR TITLE
Gc process kills itself at the end

### PR DESCRIPTION
**Pull request by icristescu**

### CODE REVIEW

File: src/irmin-pack/unix/ext.ml [hunk 1]
```diff
@@ -319,7 +319,10 @@ module Maker (Config : Conf.S) = struct
                   Gc.run_and_output_result root commit_key
                     ~generation:next_generation
                 in
-                exit 0
+                (* Once the gc is finished, the child process kills itself to
+                   avoid calling at_exit functions in upstream code. *)
+                Unix.kill (Unix.getpid ()) 9;
+                Ok ()
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/ext.ml:325]:
```suggestion
                assert false (* unreachable *);
 ```

File: src/irmin-pack/unix/ext.ml [hunk 2]
```diff
@@ -417,7 +420,7 @@ module Maker (Config : Conf.S) = struct
                   let result =
                     let open Result_syntax in
                     match (status, gc_output) with
-                    | Lwt_unix.WEXITED 0, Ok copy_end_offset ->
+                    | Lwt_unix.WSIGNALED _, Ok copy_end_offset ->
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/ext.ml:423]:
```suggestion
                    | Lwt_unix.WSIGNALED 9, Ok copy_end_offset ->
```
Does this work?

Comment (Ngoguey42) [reply src/irmin-pack/unix/ext.ml:423]: Let's unlink the result file before forking. To make sure that we are not catching the result file from a previously aborted GC

Comment (icristescu) [reply src/irmin-pack/unix/ext.ml]: > `| Lwt_unix.WSIGNALED 9,  `
> Does this work?

No, somehow it becomes a (-7) when catched here. I opted for `Sys.sigkill` instead. (changed a bit the structure as the match on  `Lwt_unix.WSIGNALED  Sys.sigkill` does not work...)

Comment (Ngoguey42) [reply src/irmin-pack/unix/ext.ml]: Nevermind, leave `_`

File: src/irmin-pack/unix/ext.ml [hunk 3]
```diff
@@ -300,6 +300,23 @@ module Maker (Config : Conf.S) = struct
...
+                match Io.unlink result_file with
+                | Error e ->
+                    [%log.warn
+                      "Unlinking temporary files from previous failed gc. \
+                       Failed with error %a"
+                      Errs.pp e]
+                | Ok () -> ())
+            | `Directory | `Other ->
+                [%log.warn "Unexpected format for %s found" result_file]
+
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/ext.ml:319]: You could have called `Io.unlink` straight away and looked at the output to print the right message. It's the _fail-fast principle_. It applies well in this situation.

Anyhow, don't bother modifying your code.

**Review by Ngoguey42 — APPROVED** [review]

File: src/irmin-pack/unix/ext.ml (hunk 3 above)
Comment (icristescu) [reply src/irmin-pack/unix/ext.ml]: I don't mind changing it :)

**Pull request merged**